File: house_robber_ii.py
class Solution:
    def rob(self,nums):
        print nums[:-1]
        return max(self.robber(nums[:-1]),self.robber(nums[1:]))
    def robber(self,nums):
        length = len(nums)
        if length < 2:
            return nums[0] if nums else 0
        dp = [0 for x in range(length)]
        dp[0] = nums[0]
        dp[1] = max(nums[0],nums[1])
        for i in range(2,length):
            dp[i] = max(dp[i-2]+nums[i],dp[i-1])
        return  dp[length-1]
    def rob1(self, nums):
        length = len(nums)
        if length < 2:
            return nums[0] if nums else 0
        dp = [0 for x in range(length)]
        dp[0] = nums[0]
        dp[1] = max(nums[0],nums[1])
        for i in range(2,length):
            dp[i] = max(dp[i-2]+nums[i],dp[i-1])
        if dp[length-3]+nums[length-1] > dp[length-2]:
            return dp[length-2]
        else:
            return dp[length-1] 


if __name__ == '__main__':
    sol = Solution()
    #nums = [1,1,1,1]
    #nums = [2,3,2]
    nums = [1,2,1,1]
    print sol.rob1(nums)

File: Algorithm/CPP/house_robber_ii.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <unordered_map>
# include <cctype>
# include <algorithm>
# include <time.h>
# include <queue>
using namespace std;

class Solution {
public:
	int rob(vector<int>& nums) {
		int n = nums.size();
		if (n == 0) return 0;
		if (n == 1) return nums[0];
		return max(rob(nums, 0, n - 1), rob(nums, 1, n));
	}
	int rob(vector<int>& nums, int start, int end) {

		int n = end - start;
		if (n == 1) return nums[start];
		vector<int> dp(n);
		dp[0] = nums[start];
		dp[1] = max(nums[start], nums[start + 1]);
		for (int i = 2; i < n; ++i)
		{
			dp[i] = max(dp[i - 2] + nums[i + start], dp[i - 1]);
		}
		return dp[n - 1];
	}
};

int main()
{
	Solution sol;
	vector<int> nums = { 4, 1, 2, 7, 5, 3, 1};
	//nums = { 1, 2 };
	int k = 8;
	//nums = { 3, 2, 1, 5, 6, 4 };
	string word1 = "vintner";
	string word2 = "writers";
	cout<< sol.rob(nums)<< endl;
	//for (auto k :nums )cout << k << " ";
	system("pause");
	return 0;
}

